### DISCUSSION TITLE
overload opertator

### DISCUSSION THREAD

**Discussion by hbadi** [pybind/pybind11#3362] [Q&A]
Hi,

I have a class A with a method operator+(A a) and also a constructor with a double parameter (defined as py::init<double> )
I 'd like to write such expression a+2.5 with an automatic cast, is it possible ?

Best regards.

**Comment by jiwaszki** [comment]
Hi, of course it is possible. It is all in docs: https://pybind11.readthedocs.io/en/stable/advanced/classes.html#operator-overloading

If you have your operator defined like this (where `_n` is some member of your `A` class):
```cpp
double operator+(A a) const { return _n + a._n; };
```
Then you are able to bind it with:
```cpp
cls.def(py::self + py:self);
cls.def(py::self + double()); // this cover your case of automatic cast
```

**Comment by hbadi** [comment]
Hi @jiwaszki 

1 - What's because the operator+ is not const, i'm using an external library where operators are not const... How can i deal with ?

2 - Moreover, given a construcor `A(double)` why the definition `cls.def(py::self + py::self);` won't automatically cast double as A (like C++ does) ?

Thanks.

**Reply by jiwaszki** [reply]
1. One way is just to use lambda, easy and clean:
```cpp
  cls.def(
      "__add__", [](A &self, A &other)
      { return self + other; },
      py::is_operator());
```
2. If you want that you **probably** need custom type caster which is more advanced, refer to this: https://pybind11.readthedocs.io/en/stable/advanced/cast/custom.html#custom-type-casters

**Comment by hbadi** [comment]
Hi @jiwaszki 

For sure, lambda helps a lot !
As i understand there's no way to write `cls.def(py::self + py::self);` with a "non-const" operarator definition `A operator+(A a) { return A(_n +a.n); };`, it's compulsory to write `A operator+(A a) const { return A(_n +a.n); };` ? event if c++ does accept it (even without -fpermissive) but not pybind11.
